src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,266 @@
+            return IRMutator::visit(op);
+        }
+
+        const std::string &producer_name = finder.producer_name;
+        const std::string &mutex_name = finder.mutex_name;
+        int tuple_size = finder.tuple_size;
+        int dimensions = finder.dimensions;
+        const char *extent_field = tuple_size == 1 ? ".extent." : ".0.extent.";
+        const char *min_field = tuple_size == 1 ? ".min." : ".0.min.";
+        const char *stride_field = tuple_size == 1 ? ".stride." : ".0.stride.";
+        Expr buffer_size = Expr(1);
+        for (int i = 0; i < dimensions; i++) {
+            Expr extent = Variable::make(Int(32), producer_name + extent_field + std::to_string(i));
+            buffer_size *= extent;
+        }
+        Expr mutex_array = Call::make(type_of<halide_mutex_array **>(),
+                                      "halide_mutex_array_create",
+                                      {buffer_size},
+                                      Call::Extern);
+        Stmt body = mutate(op->body);
+        // Allocate a scalar of halide_mutex_array.
+        // This generate halide_mutex_array *mutex[1];
+        body = Allocate::make(mutex_name,
+                              type_of<halide_mutex_array *>(),
+                              MemoryType::Stack,
+                              {},
+                              const_true(),
+                              body,
+                              mutex_array,
+                              "halide_mutex_array_destroy");
+        // Insert Let statements for the buffer's min and stride for storage flattening.
+        for (int i = 0; i < dimensions; i++) {
+            body = LetStmt::make(mutex_name + ".min." + std::to_string(i),
+                                 Variable::make(Int(32), producer_name + min_field + std::to_string(i)),
+                                 body);
+            body = LetStmt::make(mutex_name + ".stride." + std::to_string(i),
+                                 Variable::make(Int(32), producer_name + stride_field + std::to_string(i)),
+                                 body);
+        }
+
+        return ProducerConsumer::make(op->name, op->is_producer, std::move(body));
+    }
+
+    Stmt visit(const Atomic *op) override {
+        if (op->mutex_name == "") {
+            return IRMutator::visit(op);
+        }
+
+        // op->mutex_indices.size() could be 0.
+        Expr index = op->mutex_indices.size() == 1 ? op->mutex_indices[0] : Expr(0);
+        Stmt body = op->body;
+        Expr mutex = Load::make(type_of<halide_mutex_array *>(),
+                                op->mutex_name,
+                                Expr(0),
+                                Buffer<>(),
+                                Parameter(),
+                                const_true(),
+                                ModulusRemainder());
+        // Add mutex locks & unlocks
+        body = Block::make(
+            Evaluate::make(Call::make(type_of<int>(),
```

Comment:
This is not exception-safe. If the body throws an error then the mutex is never unlocked. However, halide_mutex_array_destroy *will* be called, because it's the custom free function on an Allocate node. I guess  it just has to handle the case where some of the mutexes are still locked? I guess this is OK then, but you should add a comment.

Reply:
Not sure if I fully understand this but I added some comments in https://github.com/halide/Halide/commit/19a6bd69f4e0d15ba89f597b6877bf4f2316a3af